Add cloud-native-architecture.md (#6883)
pingcap/tiflash: docs/design/2023-02-23-cloud-native-architecture.md

# The Design of TiFlash Cloud Native Architecture 
- Author: [flowbehappy](http://github.com/flowbehappy)

## Solve what problem?

The current TiFlash architecture is a typical shared nothing architecture, which brings some drawbacks:

1. Storage and computation cannot be scaled separately

> 2. The update operation and the read operation are loaded on the same nodes and affect each other, including IO, CPU and other systems resources. The task of AP query is usually heavy and has a high suddenness.

Reviewer: ```suggestion
2. Updating and reading operations are loaded on the same nodes and can affect each other, including using up IO, CPU, and other resources. AP query tasks are typically heavy and can suddenly spike in demand.
```

> 3. Slow down scaling. Scaling needs to do data synchronization between TiFlash nodes.

Reviewer: ```suggestion
3. Scaling is slow because data must be synchronized between TiFlash and TiKV nodes.
```

> 4. The cost-efficiency is not optimal. For example, there are only queries at night, but we need to keep TiFlash instances around.

Reviewer: ```suggestion
4. The cost-efficiency is not optimal. For example, TiFlash instances may need to be kept running even during periods of low query demand, even if there are only queries at night.
```

We split TiFlash node into the following nodes:

1. <b>Write Node (WN)</b> is responsible for storing raft logs, decoding the data, and writing it to memtable. The data written to disk by memtable will be uploaded to S3 for persistence. And it triggers remote compaction as appropriate. It also provides data shopshot information for query, such as data files involved in query.
2. <b>Read Node (RN)</b> is responsible for the query. To start the query, it need to first request Writes Nodes to obtain the necessary snapshot information and Memtable data in memory, and then directly request S3 to run queries, no need to interact with Write Nodes any more. It uses the local disk that comes with the ec2 node to do data caching to improve query performance.
3. <b>Compaction Worker</b> is used to handle various background tasks from Write Nodes offload to avoid affecting the foreground thread. Since it is a stateless node, it is very convenient to apply for resources dynamically and avoid compaction delays.

## Advantages
1. Storage and calculation separation reduces deployment costs
    * Storage and computing resources can be scaled separately on demand to avoid resource mismatch

>     * Light and heavy queries can be supported by different Query Tenant respectively. For example, regular heavy calculations can be temporarily started to complete the calculation and destroyed after use.

Reviewer: ```suggestion
    * Light and heavy queries can be supported by different Query Tenants respectively. For example, heavy computations can be temporarily launched to complete calculations and then destroyed after use.
```

> 2. Read and write separation improves stability. Synchronize data, organize logic, and query load are physical separated, which can avoid heavy AP query load crowding out system resources.

Reviewer: ```suggestion
2. Separating reads and writes improves stability. Data synchronization, data reorganization logic, and query load are physically separated, which can prevent heavy AP query loads from crowding out system resources.
```

3. Greatly increase the speed of volume expansion and contraction
    * Because RN is stateless, in theory, expansion and Load Balance can be completed in a few seconds.

>     * The Region migration on WN does not need to resynchronize the snapshot from the TiKV , and the migration is completed directly by downloading or referencing the S3 file. It is expected that a new node can complete the Load Balance within 1min.

Reviewer: ```suggestion
    * The Region migration on WN does not need to resynchronize the snapshot from the TiKV. Instead, migration is completed by downloading or referencing the S3 file directly. It is expected that a new node can complete load balancing within one minute.
```

> 4. Improves Scalability while retaining the performance benefits of Locality
>     * RN can use the local disk as LRU cache, so that most of the hot data does not need to read S3

Reviewer: ```suggestion
4. Improves Scalability while retaining the performance benefits of locality
    * RN can use the local disk as an LRU cache, so that most of the hot data does not need to read S3
```

2. It supports for External Pages, that is, storing larger Pages externally. This feature allows PS to manage large files, such as DTFile (SSTFile similar to LSM Tree).
3. It supports for Ref Pages, similar to hard link for file system
4. It supports MVCC Snapshot, that is, the application layer can obtain a fixed data view according to the Snapshot

Managing these complex IO dependencies at the application layer is doomed to have no future. We choose to store all TiFlash data in PageStorage (PS), and then PS upload all data changes regularly , which can solve the consistency problem.

It works like EBS snapshot, i.e. uploading changes to S3 periodically.

1. It has been guaranteed that if the PS state on the TiFlash disk at a certain time t is S(t), at this time restart, the PS based on the local disk can be restored to the state S(t)
2. If we get PS snapshot at time t and upload all the data in the snapshot to S3 to form file F, then the TiFlash state recovered from F must be equal to S(t)

Here is the new architecture:

![](images/2023-02-23-cloud-native-architecture-3.png)

## Key Implementation Points

> 

Author: ```suggestion

* Periodically upload checkpoints and delta data to S3 by WN.
```

> * Store raft logs data into PageStorage. Currently it is stored in [raft-engine](https://github.com/tikv/raft-engine).

Author: * Periodically upload checkpoints and delta data to S3 by WN.

3. It supports for Ref Pages, similar to hard link for file system
4. It supports MVCC Snapshot, that is, the application layer can obtain a fixed data view according to the Snapshot

Managing these complex IO dependencies at the application layer is doomed to have no future. We choose to store all TiFlash data in PageStorage (PS), and then PS upload all data changes regularly , which can solve the consistency problem.

It works like EBS snapshot, i.e. uploading changes to S3 periodically.

1. It has been guaranteed that if the PS state on the TiFlash disk at a certain time t is S(t), at this time restart, the PS based on the local disk can be restored to the state S(t)
2. If we get PS snapshot at time t and upload all the data in the snapshot to S3 to form file F, then the TiFlash state recovered from F must be equal to S(t)

Here is the new architecture:

![](images/2023-02-23-cloud-native-architecture-3.png)

## Key Implementation Points

* Periodically upload checkpoints and delta data to S3 by WN.
* Store raft logs data into PageStorage. Currently it is stored in [raft-engine](https://github.com/tikv/raft-engine).

> * Support generating query snapshots by WN and sending latest data to RN

Reviewer 2: ```suggestion
* Support generating query snapshots by WN and sending latest data to RN.
```
